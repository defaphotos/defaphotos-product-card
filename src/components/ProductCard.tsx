import React, { createContext } from 'react';
import styles from '../styles/styles.module.css';
import useProducts from '../hooks/useProducts';
import { ContextProductProps, ProductCardProps } from '../interfaces/interfaces';

export const ContextProduct = createContext({} as ContextProductProps);
const { Provider } = ContextProduct;

export const ProductCard = ({product, children, className, style, onChange, value, initialValues}:ProductCardProps) => {

    const {counter, increaseBy, maxCount, isMaxCountReached, reset} = useProducts({onChange, product, value, initialValues});

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount,
        }}>
            <div style={style} className={`${styles.productCard} ${className}`}>
                {children({count: counter, increaseBy, isMaxCountReached, product, reset, maxCount })} 
            </div>
        </Provider>
    );
};