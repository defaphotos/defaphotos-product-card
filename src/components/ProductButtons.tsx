import React, { useCallback, useContext } from "react";
import { ContextProduct } from "./ProductCard";
import styles from '../styles/styles.module.css';
import { ProductButtonsProps } from "../interfaces/interfaces";

export const ProductButtons = ({ className, style }: ProductButtonsProps) => {

    const { counter, increaseBy, maxCount } = useContext(ContextProduct);

    const isMaxReached = useCallback(
      () => {
        console.log('useCallback', counter)
        return !!maxCount && counter === maxCount
      },
      [counter, maxCount],
    )
    
    return (
        <div style={style} className={`${styles.buttonsContainer} ${className}`}>
            <button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>-</button>

            <div className={styles.countLabel}>{counter}</div>

            <button onClick={() => increaseBy(1)} className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}>+</button>
        </div>
    )
}