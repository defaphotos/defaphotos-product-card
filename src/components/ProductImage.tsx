/* eslint-disable jsx-a11y/alt-text */
import React, { useContext } from "react";
import noImage from '../assets/no-image.jpg'
import { ProductImageProps } from "../interfaces/interfaces";
import styles from '../styles/styles.module.css';
import { ContextProduct } from "./ProductCard";

export const ProductImage = ({img, className, style}: ProductImageProps) => {

    const { product } = useContext(ContextProduct);
    let showToImg: string;
    if (img) {
        showToImg = img;
    } else if (product.img) {
        showToImg = product.img;
    } else {
        showToImg = noImage;
    }
    
    return (
        <img style={style} className={`${styles.productImg} ${className}`} src={showToImg} /> 
    )
}