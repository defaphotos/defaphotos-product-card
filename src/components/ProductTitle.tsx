import React, { useContext } from "react"
import { ContextProduct } from "./ProductCard"
import styles from '../styles/styles.module.css';
import { ProductTitleProps } from '../interfaces/interfaces';

export const ProductTitle = ({className, title, style}: ProductTitleProps) => {

    const {product} = useContext(ContextProduct);
    let showTitle;

    if (title) {
        showTitle = title;
    } else if (product.title) {
        showTitle = product.title;
    }

    return (
        <span style={style} className={`${styles.productDescription} ${className}`}>{showTitle}</span>
    )
}