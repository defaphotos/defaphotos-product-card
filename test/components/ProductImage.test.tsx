import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import {product2} from '../data/products';
import { ProductImage } from '../../src/components/ProductImage';

describe('ProductImage', () => {
    test('Debe mostrar el componente con la producto sin imagen', () => {
        const wrapper = renderer.create(
            <ProductImage img="https://hola.jpg" />
        )
        expect(wrapper.toJSON()).toMatchSnapshot()
    })


    test('Debe mostrar el componente con la producto con imagen', () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })
})