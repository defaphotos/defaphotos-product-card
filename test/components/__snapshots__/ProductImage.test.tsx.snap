// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente con la producto con imagen 1`] = `
<div
  className="productCard undefined"
>
  <img
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage Debe mostrar el componente con la producto sin imagen 1`] = `
<img
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;
